File: Pandacap/Podcasts/PodcastStreamProvider.cs
using System.Net.Http.Headers;

namespace Pandacap.Podcasts
{
    public class PodcastStreamProvider(
        IHttpClientFactory httpClientFactory)
    {
        public async Task<PodcastAccessor> CreateAccessorAsync(Uri uri, CancellationToken cancellationToken)
        {
            using var client = httpClientFactory.CreateClient();

            using var headRequest = new HttpRequestMessage(HttpMethod.Head, uri);
            using var headResponse = await client.SendAsync(headRequest, cancellationToken);

            headResponse.EnsureSuccessStatusCode();

            if (headResponse.Content.Headers.ContentLength is long length)
                return new()
                {
                    Stream = new BufferedStream(
                        new PodcastStream(
                            httpClientFactory,
                            uri,
                            length),
                        bufferSize: 1048576),
                    ContentLength = headResponse.Content.Headers.ContentLength,
                    ContentType = headResponse.Content.Headers.ContentType
                };

            using var request = new HttpRequestMessage(HttpMethod.Get, uri);

            var response = await client.SendAsync(
                request,
                HttpCompletionOption.ResponseHeadersRead,
                cancellationToken);

            return new()
            {
                Stream = await response.Content.ReadAsStreamAsync(cancellationToken),
                ContentLength = response.Content.Headers.ContentLength,
                ContentType = response.Content.Headers.ContentType
            };
        }
    }
}

File: Pandacap.Podcasts/PodcastStreamProvider.fs
namespace Pandacap.Podcasts

open System
open System.IO
open System.Net.Http
open System.Threading

type PodcastStreamProvider(httpClientFactory: IHttpClientFactory) =
    let initPodcastStream length uri: Stream =
        new PodcastStream(httpClientFactory, uri, length)

    let buffer bufferSize stream: Stream =
        new BufferedStream(stream, bufferSize)

    member _.CreateAccessorAsync(uri: Uri, cancellationToken: CancellationToken) = task {
        use client = httpClientFactory.CreateClient()

        use headRequest = new HttpRequestMessage(HttpMethod.Head, uri)
        use! headResponse = client.SendAsync(headRequest, cancellationToken)

        match headResponse.EnsureSuccessStatusCode().Content.Headers.ContentLength |> Option.ofNullable with
        | Some length ->
            return {|
                Stream =
                    uri
                    |> initPodcastStream length
                    |> buffer 1048576
                ContentType = headResponse.Content.Headers.ContentType
                EnableRangeProcessing = true
            |}
        | None ->
            use request = new HttpRequestMessage(HttpMethod.Get, uri)

            let! response = client.SendAsync(request, cancellationToken)
            let! stream = response.EnsureSuccessStatusCode().Content.ReadAsStreamAsync(cancellationToken)

            return {|
                Stream = stream
                ContentType = response.Content.Headers.ContentType
                EnableRangeProcessing = false
            |}
    }
